core/src/main/java/com/microsoft/applicationinsights/LowLevelClient.java

```diff
@@ -0,0 +1,119 @@
+package com.microsoft.applicationinsights;
+
+import com.azure.core.http.*;
+import com.azure.core.http.policy.HttpPipelinePolicy;
+import com.azure.core.util.tracing.Tracer;
+import com.azure.monitor.opentelemetry.exporter.implementation.models.TelemetryItem;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.microsoft.applicationinsights.internal.authentication.AadAuthentication;
+import io.opentelemetry.sdk.common.CompletableResultCode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Flux;
+import reactor.util.context.Context;
+
+import java.io.IOException;
+import java.net.URL;
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+import java.util.List;
+
+// TODO performance testing
+class LowLevelClient {
+
+    private static final Logger logger = LoggerFactory.getLogger(LowLevelClient.class);
+
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    private static final ByteBufferPool byteBufferPool = new ByteBufferPool();
+
+    static {
+        mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
+    }
+
+    private final HttpPipeline pipeline;
+    private final URL endpoint;
+
+    LowLevelClient(URL endpoint) {
+        HttpClient client = HttpClient.createDefault();
+        HttpPipelineBuilder pipeline = new HttpPipelineBuilder()
+                .httpClient(client);
+        // TODO handle authentication exceptions
+        HttpPipelinePolicy authenticationPolicy = AadAuthentication.getInstance().getAuthenticationPolicy();
+        if (authenticationPolicy != null) {
+            pipeline.policies(authenticationPolicy);
+        }
+        // TODO check existing retry policy, and its configuration
+        this.pipeline = pipeline.build();
+        this.endpoint = endpoint;
+    }
+
+    CompletableResultCode send(List<TelemetryItem> telemetryItems) {
+        List<ByteBuffer> byteBuffers;
+        try {
+            byteBuffers = encode(telemetryItems);
+        } catch (Throwable t) {
+            logger.error("Error encoding telemetry items: {}", t.getMessage(), t);
+            return CompletableResultCode.ofFailure();
+        }
+        try {
+            return internalSend(byteBuffers);
+        } catch (Throwable t) {
+            logger.error("Error sending telemetry items: {}", t.getMessage(), t);
+            return CompletableResultCode.ofFailure();
+        }
+    }
+
+    List<ByteBuffer> encode(List<TelemetryItem> telemetryItems) throws IOException {
+        ByteBufferOutputStream out = new ByteBufferOutputStream(byteBufferPool);
+        for (Iterator<TelemetryItem> i = telemetryItems.iterator(); i.hasNext(); ) {
+            mapper.writeValue(out, i.next());
+            if (i.hasNext()) {
+                out.write('\n');
+            }
+        }
+        out.close(); // closing ByteBufferOutputStream is not really needed, but this makes LGTM happy
```

Comment:
it would be nice if calling close() in a finally block in the case of exception, it will get close properly.  is it ok to not make this throwing an IOException and handle exception here instead of line 54 - 59?  if there is an exception thrown from line 69 - 75, bytebufferoutputstream is not properly closed yet.

Reply:
clarified comment, and returned buffers to pool on exception